src/libopensc/card-piv.c

```diff
@@ -157,11 +178,218 @@ enum {
+static cipher_suite_t css[PIV_CSS_SIZE] = {
+		{PIV_CS_CS2, 256, NID_X9_62_prime256v1, {{1, 2, 840, 10045, 3, 1, 7, -1}},
+		PIV_CS_CS2, 65, 16, 32, 61,
+		4, 0x09, 8, 1, 16, 8, 16, 1,
+		4, 128/8, SHA256_DIGEST_LENGTH,
+		(EVP_MD *(*)(void)) EVP_sha256,
+		(const EVP_CIPHER *(*)(void)) EVP_aes_128_cbc,
+		(const EVP_CIPHER *(*)(void)) EVP_aes_128_ecb,
+		"aes-128-cbc", "aes-128-ecb",
+		"prime256v1"},
+
+		{PIV_CS_CS7, 384, NID_secp384r1, {{1, 3, 132, 0, 34, -1}},
+		PIV_CS_CS7, 97, 16, 48, 69,
+		4, 0x0D, 8, 1, 16, 8, 24, 1,
+		4, 256/8, SHA384_DIGEST_LENGTH,
+		(EVP_MD *(*)(void)) EVP_sha384,
+		(const EVP_CIPHER *(*)(void)) EVP_aes_256_cbc,
+		(const EVP_CIPHER *(*)(void)) EVP_aes_256_ecb,
+		"aes-256-cbc", "aes-256-ecb",
+		"secp384r1"}
+	};
+// clang-format on
+
+/* 800-73-4  4.1.5 Card Verifiable Certificates */
+typedef struct piv_cvc {
+	sc_pkcs15_der_t der;					// Previous read der
+	int cpi;						// Certificate profile indicator (0x80)
+	char issuerID[8];					// Issuer Identification Number
+	size_t issuerIDlen;					//  8 bytes of sha-1 or 16 byte for GUID
+	u8  subjectID[16];					//  Subject Identifier (8) or GUID (16)  == CHUI
+	size_t subjectIDlen;					//  8 bytes of sha-1 or 16 byte for GUID
+	struct sc_object_id pubKeyOID;				// Public key algorithm object identifier
+	u8 *publicPoint;					// Public point for ECC
+	size_t publicPointlen;
+	int roleID;						// Role Identifier 0x00 or 0x12
+	u8 *body;						// signed part of CVC in DER
+	size_t bodylen;
+	struct sc_object_id  signatureAlgOID;			// Signature Algroithm Identifier
+	u8 *signature;						// Certificate signature DER
+	size_t signaturelen;
+} piv_cvc_t;
+
+#define PIV_SM_MAX_FIELD_LENGTH  384
+#define PIV_SM_MAX_MD_LENGTH	SHA384_DIGEST_LENGTH
+
+#define PIV_SM_FLAGS_SM_CERT_SIGNER_VERIFIED	0x000000001lu
+#define PIV_SM_FLAGS_SM_CVC_VERIFIED		0X000000002lu
+#define PIV_SM_FLAGS_SM_IN_CVC_VERIFIED		0x000000004lu
+#define PIV_SM_FLAGS_SM_CERT_SIGNER_PRESENT	0x000000010lu
+#define PIV_SM_FLAGS_SM_CVC_PRESENT		0X000000020lu
+#define PIV_SM_FLAGS_SM_IN_CVC_PRESENT		0x000000040lu
+#define PIV_SM_FLAGS_SM_IS_ACTIVE		0x000000080lu	/* SM has been started */
+	/* if card supports SP800-73-4 SM: */
+#define PIV_SM_FLAGS_NEVER			0x000000100lu	/* Don't use SM even if card support it */
+								/* Default is use if card supports it */
+								/* will use VCI if card supports it for contactless */
+#define PIV_SM_FLAGS_ALWAYS			0x000000200lu	/* Use SM or quit, VCI requires SM */
+#define PIV_SM_FLAGS_DEFER_OPEN			0x000001000lu	/* call sm_open from reader_lock_obtained */
+#define PIV_SM_VCI_ACTIVE			0x000002000lu   /* VCI is active */
+#define PIV_SM_GET_DATA_IN_CLEAR		0x000004000lu	/* OK to do this GET DATA in the clear */
+
+typedef struct piv_sm_session {
+	/* set by piv_sm_open */
+	int aes_size; /* 128 or 256 */
+
+	u8 SKcfrm[32];
+	u8 SKmac[32];
+	u8 SKenc[32];  /* keys are either AES 128 or AES 256 */
+	u8 SKrmac[32];
+	u8 enc_counter[16];
+	u8 enc_counter_last[16];
```

Comment:
Are you referring to the `SK*[32]` variables or the enc_counter?

If you are referring to the SK variables, (OpenSSL  < 3.0)  `CMAC_Init`  and (openSSL >= 3.0) `EVP_MAC_init` take these as  an array of unsigned bytes and a length. In OpenSSL 3.0 all symmetric keys are treated as arrays of unsigned bytes and  lengths.  So convert to BIGNUM does not accomplish anything.

If enc_counter it could be possible to use: BN_bn2bin  and BN_bin2bn to increment.

Reply:
yes, I'm referring to the encryption/decryption counters. BIGNUM lets you avoid managing this byte per byte

Reply:
I don't think it is worth converting. The counters are passed as 16 bytes in a number of places and the piv_inc increments the counters:   enc_counter and dec_counter.   800-74-4: 
"The initial value of the encryption counter upon successful completion of the key establishment protocol shall be '00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01'. The encryption counter shall be incremented by one after each APDU sent over secure messaging"  and  "The encryption counter value used to generate the IV to encrypt the response data shall be the same as the encryption counter value used to generate the IV to encrypt the corresponding request data, with the exception that the most significant byte of the 16-byte counter shall be set to '80' (i.e., the IV used to encrypt the first response after successful completion of the key establishment protocol shall be generated by encrypting '80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01' with SKENC"

Reply:
Sure, I have read the code and the spec. The idea was to make the code more readable and to avoid piv_inc(). size_t would also be a possible option, but if you prefer using a buffer it's OK for me.